### ISSUE TITLE
[-Wdangling-gsl] improve detect more compilcate codes, pls.

### ISSUE THREAD

**Issue by raidenluikang** [llvm/llvm-project#60851]
code: https://godbolt.org/z/oE9f3s71h
```

#include <stddef.h>
#include <utility>
#include <array>
#include <vector>
#include <iterator>
#include <string>
 
#include <stdio.h>

struct abc
{
	std::string str;
	std::string_view view() const { return str; };
};
std::string foo(){
	return "jhfdsljfldskjfl;ksjdfsfda;ljfsdlakjfkl;sajdf;lsadjfk;lsajdf;lajf";
} 


int main()
{
std::string_view abc_view = abc{"dslfkjsladfjfsdkjhfklsdfl;sjdfl;jsdlf;jsldk;f;lsdfj"}.view(); // --> Expected there should be warning , also.
    std::string_view foo_view = foo();
	printf("abc_view = %s\n", abc_view.data());
    printf("foo_view = %s\n", foo_view.data());
}
```

Warnings:

```
<source>:24:33: warning: object backing the pointer will be destroyed at the end of the full-expression [-Wdangling-gsl]
    std::string_view foo_view = foo();
                                ^~~~~
1 warning generated.
ASM generation compiler returned: 0
<source>:24:33: warning: object backing the pointer will be destroyed at the end of the full-expression [-Wdangling-gsl]
    std::string_view foo_view = foo();
                                ^~~~~
1 warning generated.
Execution build compiler returned: 0
Program returned: 0
abc_view = 
foo_view = 

```

**Comment by raidenluikang** [comment]
Another similar code:


```
#include <stddef.h>
#include <utility>
#include <array>
#include <vector>
#include <iterator>
#include <string>
 
#include <stdio.h>

struct abc
{
	std::string str;
	std::string_view view() const { return str; };
};
std::string foo(){
	return "jhfdsljfldskjfl;ksjdfsfda;ljfsdlakjfkl;sajdf;lsadjfk;lsajdf;lajf";
} 


int main()
{
	std::string_view abc_view = abc{"dslfkjsladfjfsdkjhfklsdfl;sjdfl;jsdlf;jsldk;f;lsdfj"}.view();//expected there should be warning
    
    std::string_view foo_view = foo();
    
    std::string str = "dslkfjsl;akdjfkl;sdjfl;ksdjfkljsdla;kfjlsad;kfjk;lsadjfksajdfk;lajf;l";

    std::string_view substr_view {str.substr(0,5)}; //Expected there should be warning
	
    printf("abc_view = %s\n", abc_view.data());
    
    printf("foo_view = %s\n", foo_view.data());

    printf("str_view = %s\n", substr_view.data());
}

```

**Comment by royjacobson** [comment]
You might want to take a look at [[clang::lifetimebound]]

**Comment by raidenluikang** [comment]
> You might want to take a look at [[clang::lifetimebound]]

A quick search it, says that it need put some function/method declaration.

How detect existing class situation?

```
std::string str = "hello world";

.....  a lot of code ..


std::string_view view = str.substr(0, 5); // where I should put [[clang::lifetimebound]] ?
``

**Comment by frederick-vs-ja** [comment]
See https://github.com/llvm/llvm-project/issues/62390#issuecomment-1589524195.

**Comment by usx95** [comment]
`lifetimebound` should be added to the `view` on `this` param. https://godbolt.org/z/7TKnnMnW5
